// wikipediajs/dist/generators/index.js
"use strict";

var _require = require("../constants/index"),
    QUERY_GENERATOR = _require.QUERY_GENERATOR,
    RANDOM_GENERATOR = _require.RANDOM_GENERATOR;

var _require2 = require("../utils/urlUtils"),
    queryApi = _require2.queryApi;

var DEFAULT_PARAMS = {
  prop: "info",
  inprop: "url",
  format: "json",
  action: "query"
};

function doSearch(searchTerm) {
  var lang = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "en";
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  var params = _getParams(options, {
    gsrsearch: searchTerm,
    generator: QUERY_GENERATOR
  });

  return queryApi(lang, params).then(function (response) {
    return response;
  });
}

function doRandom() {
  var lang = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "en";
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var params = _getParams(options, {
    generator: RANDOM_GENERATOR
  });

  return queryApi(lang, params).then(function (response) {
    return response;
  });
}

function doAllCatergories() {
  var lang = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "en";
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var params = _getParams(options, {
    list: "allcategories"
  });

  return queryApi(lang, params).then(function (response) {
    return response;
  });
}

function _getParams() {
  var queryOptions = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var generatorOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  return Object.assign({}, DEFAULT_PARAMS, queryOptions, generatorOptions);
}
module.exports = {
  doSearch: doSearch,
  doRandom: doRandom,
  doAllCatergories: doAllCatergories
};
